contracts/contracts/BLSExpander.sol

```diff
@@ -145,4 +154,42 @@ contract BLSExpander {
     //     );
     // }
 
-}
\ No newline at end of file
+    function registerPublicKey(
+        address walletAddress,
+        uint256[4] calldata publicKey
+    ) public {
+        addressToPublicKey[address(walletAddress)] = publicKey;
+    }
+
+    function registerPublicKeys(
+        address[] calldata walletAddresses,
+        uint256[4][] calldata publicKeys
+    ) external {
+        for (uint256 i=0; i<publicKeys.length; i++) {
+            registerPublicKey(walletAddresses[i], publicKeys[i]);
+        }
+    }
+
+    function addressProcessBundle(
+        AddressBundle memory addressBundle
+    ) external returns (
+        bool[] memory successes,
+        bytes[][] memory results
+    ) {
+        // Expand addresses to public keys
+        uint256 numPubKeys = addressBundle.senderAddresses.length;
+        uint256[4][] memory senderPublicKeys =
+            new uint256[4][](numPubKeys);
+        for (uint256 i=0; i<numPubKeys; i++) {
+            senderPublicKeys[i] = addressToPublicKey[addressBundle.senderAddresses[i]];
+        }
+
+        // Use them to re-create bundle
+        VerificationGateway.Bundle memory bundle;
+        bundle.signature = addressBundle.signature;
+        bundle.senderPublicKeys = senderPublicKeys;
+        bundle.operations = addressBundle.operations;
+
+        return verificationGateway.processBundle(bundle);
+    }
```

Comment:
@jzaki I assume these changes are ok to go in since the expander is not a core contract. Imagine we will carry these optimizations further in the future.

Reply:
Confirmed.